### ISSUE TITLE
Add form validation for create/edit views

### ISSUE THREAD

**Issue by jeancochrane** [fpdcc/document-search#41]
We've shied away from form validation so far because the data quality wasn't good enough, but now that active users are starting to use the application we should add it in to make sure that new data meets the standard.

For Add views, we can drop this validation in easily. For edit views, we'll need a custom validator to check if there's pre-existing data for the record and skip the validation if so. (Remember to turn off JavaScript validation in these cases.)

**Comment by jeancochrane** [comment]
I've got a preliminary spike of the form validation framework that seems to work well for licenses. Now we just need a canonical list of validators for each field and I can roll it out across all the document types. Here's my guess of what validators should look like, based on the existing data:

## All

- **Source file:** .pdf extension

## Books

- **Township:** Integer range between 35-42
- **Range:** Integer range between 9-15
- **Section:** Integer range between 1-36

## Control Monument Maps

- **Township:** Integer between 35-42
- **Range:** Integer between 9-15
- **Section:** Array of integers between 1-36
- **Part of section:** Choice of "E1/2" or "W1/2"

## Surplus Parcels

- **Surplus parcel:** No restrictions
- **Description:** No restrictions

## Deep Tunnels:

- **Descrpition:** No restrictions

## Dossiers

- **File number:** Positive integer
- **Document number:** Positive integer

## Easements

- **Easement number:** Positive integer
- **Description:** No restrictions

## Flat Drawings

- **Area:** Integer between 1-33
- **Section:** Integer range between 1-36
- **Map number:** No restrictions
- **Location:** No restrictions
- **Building ID:** No restrictions
- **Description:** No restrictions
- **Job number:** No restrictions
- **Number of sheets:** Positive integer
- **Date:** Date formatted like "YYYY-MM-DD"
- **Cross reference area:** Integer between 1-33
- **Cross reference section:** Integer between 1-36
- **Cross reference map number:** No restrictions
- **Hash:** No restrictions
- **CAD file:** One of:
    - .DWG
    - .DXF
    - .DGN
    - .STL

## Index Cards

- **Monument number:** No restrictions
- **Township:** Integer between 35-42
- **Section:** Integer range between 1-36
- **Corner:** No restrictions

## Licenses

- **Township:** Array of integers between 35-42
- **Range:** Array of integers between 9-15
- **Section:** Array of integers between 1-36
- **License number:** Hyphenated string starting with "O" and ending with an integer, like "O-100"
- **Description:** No restrictions
- **Geometry:** GeoJSON object
- **Type:** One of:
    - combined sewer
    - electric
    - gas
    - other
    - pipeline
    - sanitary sewer
    - storm sewer
    - telecom
    - water main
- **Entity:** No restrictions
- **Diameter:** Positive integer
- **Material:** No restrictions
- **End date:** Date formatted like "YYYY-MM-DD"
- **Status:** One of:
    - TBD
    - active
    - cancelled
    - continuous
    - expired
    - indefinite
    - perpetual
- **Agreement type:** No restrictions

## Project files

- **Area:** Integer between 1-33
- **Section:** Integer between 1-36
- **Job number:** No restrictions
- **Job name:** No restrictions
- **Description:** No restrictions
- **Cabinet number:** Integer between 1-10
- **Drawer number:** No restrictions

## Rights of way

- **Folder tab:** No restrictions

## Surveys

- **Township:** Array of integers between 35-42
- **Range:** Array of integers between 9-15
- **Section:** Array of integers between 1-36
- **Map number:** No restrictions
- **Location:** No restrictions
- **Description:** No restrictions
- **Job number:** 0-padded three-digit integer
- **Number of sheets:** Positive integer
- **Date:** Date formatted like "YYYY-MM-DD"
- **Cross reference area:** Integer between 1-33
- **Cross reference area:** Integer between 1-33
- **Cross reference section:** Integer between 1-36
- **Cross reference map number:** No restrictions
- **Hash:** No restrictions

## Titles

- **Control number:** Positive integer

**Comment by smcalilly** [comment]
locations (flat drawing) and dates (todo: find model). relate to the related issues @smcalilly

**Comment by xmedr** [comment]
Found some discrepancies with the data we have and the validation requirements listed here, so I just wanted to double check that these were still what we're going for or if some need to be re-evaluated.

### Flat Drawings and Index Cards
- Both have a `section` field that's listed as needing to be two integers within a range of 1 through 36. All the inputs I've seen so far have been a single number as opposed to something like "20 - 30" and they go as high as 98. 
  - By contrast, the `section` field for Books is storing that data this way just fine, and even has a custom field type set up for it that is not being used for flatdrawings/indexcards

### Index Cards
- The `township` field is listed as needing to be an integer between 35 and 42. The data is only ever the names of the township in all caps, of which there are 17 different names.
  - By contrast, all other documents' township fields are numbers.

### Surveys
- The `job_number` field is listed as needing to be a 0-padded, three digit number. The data instead has been anywhere from a 3 digit number, to a 5 digit, to a range, to just the word "Hold"

**Comment by smcalilly** [comment]
We're going to validate the fields that we have no questions and revisit the other fields in another PR.

**Comment by antidipyramid** [comment]
Let's look into:
- adding validation specifically for the flat drawing add page
- query for a list of flat drawings with incorrect section values

**Comment by xmedr** [comment]
@jim-z I'm working on adding validation to flatdrawing sections, and wanted to pass along this record of the flatdrawings within the production site that contain section values larger than 36. There are 106 invalid sections out of 7961 flatdrawings.

[faulty_flatdrawing_sections.csv](https://github.com/user-attachments/files/30990784/faulty_flatdrawing_sections.csv)

As per discussion during out last meeting, I'll be making it so flatdrawing sections are a range of two numbers between 1 and 36, inclusive. All the current values are single numbers, so I was planning to default those a range with the same number twice (i.e. a section of `14` would be turned into a range of `14 - 14`). If we should be doing something else with those defaults, I can definitely adjust. I can also do the same thing to index card sections if time permits.

Let me know if there's any other investigations you'd like for us to make into the current data.

**Comment by xmedr** [comment]
@jim-z A couple of questions/comments for you as I adjust the validation of FlatDrawing sections. A few of them have a section of `0`. In those cases, would it make sense to set them to just a `null` value, or to a default like `1 - 1`?

---

Also I'll likely skip over fixing faulty sections on IndexCards for now, since those have a bit more variance. Those are just free form strings as opposed to being restricted to only integer. Because of that, It looks like 40 of the ~2400 of them have one of the following values:
- `36-42-12`
- `5-36-12`
- `36-4`
- `7-37-12`
- `37-12`

**Comment by jim-z** [comment]
@xmedr please set section to null for those currently marked 0 in flatdrawings.  That will be a good cue for us to go back and populate the correct section number.

Apologies, I might have missed this discussion at beginning of our last meeting but I'm a little unclear on why we are shifting to storing section numbers as an inclusive range.  I think there is a need to store multiple section numbers but they wont always be numbers that are right next to each other... For instance, a flat file drawing could represent three distinct locations we would ideally want to tag: area 1/section 1, area1/section16, area 4/section 1. In this case, it might make more sense to create some sort of array field for storing all possible area/section combinations (1-1, 1-16, 4-1). See issue #140 for more on this.

**Comment by jim-z** [comment]
I don't know if this helps but to provide more context, the way I think of area and section is sort of like latitude and longitude....it is just a way to assign a physical on the ground location to a document. Having section number without area number doesn't really help narrow location down, just as you would need both latitude and longitude to define an exact point on earth.

**Comment by xmedr** [comment]
Oh interesting! We should definitely chat about this a bit more next time we meet. I'll hold off on making the change to turn them into ranges.

You're right in that at the beginning of our meeting last time, we had come to conclusion that the section should be a range, but we definitely don't have to do that. Instead, it sounds like the specific pairs of numbers is more what's useful to you all, is that right? In that case, it makes sense as to why they're currently single numbers, especially if the same flat drawing is duplicated a number of times to represent the different area/section combinations.

If so, maybe what would be useful is a kind of single `area/section` field, that stores multiple sets of two numbers. We could make that field, and combine the existing values of area + section - in that order - to populate them. Then we could do away with the old individual area/section fields.

We may even be able to combine some flat drawing objects, provided there's some way to identify that they're the same. We would also need to reconcile some of them having different properties in other fields if that's the case. Just a preliminary thought, I'll likely have more questions as we work through this.

Also, to be clear, is "flat file" another name for a "flat drawing"?

**Comment by jim-z** [comment]
Correct, it is the area/section combination that is most important to us as a primary document location identifier. The legacy "cross ref section" and "Cross ref area" fields were another work around used for storing multiple area/section combinations before the files got digitized...but even then they could only record two combinations max. I know this might be a big lift with implications for how the UI functions but I think it would be a major improvement to be able to assign locations more accurately/efficiently. We will discuss more at our next meeting!

"Flat file" is a reference to the flat drawer filing cabinet where the documents in that bucket were originally stored.  Yes, they are mostly all oversized design drawings/blue prints for our critical infrastructure.  There is some other stuff mixed in there like random historical maps, surveys, etc but they are all oversized drawing/map documents...it's taken me years to decipher the original organization system that this is all based on and I'm still finding new surprises every day :(

**Comment by antidipyramid** [comment]
Paused in favor of #142
